pkg/pushsync/pushsync.go

```diff
@@ -77,59 +78,153 @@ func (ps *PushSync) Close() error {
 	return nil
 }
 
-// handler handles chunk delivery from other node and inserts it to localstore.
-// it also sends this chunk to the closest peer if one exists.
+// handler handles chunk delivery from other node and forwards to its destination node.
+// If the current node is the destination, it stores in the local store and sends a receipt.
 func (ps *PushSync) handler(ctx context.Context, p p2p.Peer, stream p2p.Stream) error {
-	_, r := protobuf.NewWriterAndReader(stream)
+	w, r := protobuf.NewWriterAndReader(stream)
 	defer stream.Close()
 
-	var ch pb.Delivery
-
-	if err := r.ReadMsg(&ch); err != nil {
-		if err == io.EOF {
-			return nil
-		}
-		return err
+	// Get the delivery
+	ch, err := ps.getChunkDelivery(r)
```

Comment:
you could return the chunk directly from `getChunkDelivery`

Reply:
What if we get error reading from the stream or decoding the protobuf?

Reply:
I meant that instead of returning the protobuf delivery message you can return the chunk. I did not say not to return an error

Reply:
aaha.. ok.. my bad..